### PR TITLE
Align axes and parametric functions/surfaces when using `Axes` and `ThreeDAxes`

### LINKED ISSUE #1446: z-axis and parametric surface misplaced in ThreeDAxes
## Description of bug / unexpected behavior

When using `ThreeDAxes` in v0.6.0 to create a plot where the minimum values of the x-axis and y-axis are not 0, the z-axis is not adjusted to go through where the x-axis and y-axis cross.  Meanwhile, any parametric surface added to the plot is also not moved to account for the shift in the x-axis and y-axis.  Here is an example:

https://user-images.githubusercontent.com/79891353/116828508-0c7c3000-ab54-11eb-9bb6-03411961d5df.mp4

## Expected behavior

The z-axis should automatically adjust, just as the x-axis and y-axis both adjust according to the `x_range` and `y_range` variables.  Likewise, when adding a parametric surface, or any `mobject` to the axes, it too should be adjusted to the correct position.

## How to reproduce the issue

<details><summary>Code for reproducing the problem</summary>

```py
from manim import *
import numpy as np

class ThreeDSurfacePlot(ThreeDScene):
    def construct(self):
        resolution_fa = 42
        self.set_camera_orientation(phi=75 * DEGREES, theta=10 * DEGREES)
        
        axes = ThreeDAxes(
                x_range=(5, 10, 1),
                y_range=(3, 8, 1),
                z_range=(-1, 1, 0.5)
                )

        def param_surface(u, v):
            x = u
            y = v
            z = np.sin(x) * np.cos(y)
            return z

        surface_plane = ParametricSurface(
            lambda u, v: axes.c2p(u, v, param_surface(u, v)),
            resolution=(resolution_fa, resolution_fa),
            v_min=5,
            v_max=10,
            u_min=3,
            u_max=8,
        )

        surface_plane.set_style(fill_opacity=1, stroke_color=GREEN)
        surface_plane.set_fill_by_checkerboard(ORANGE, BLUE, opacity=0.5)
                                    
        self.add(axes, surface_plane)
        
        self.begin_ambient_camera_rotation(rate=-0.3)
        self.wait(10)
```

</details>

## System specifications

<details><summary>System Details</summary>
Manim v0.6.0
</details>

## Additional comments

I will try to investigate this issue and report back my findings.

### PR DESCRIPTION
**Pull request by alembcke**
## Changelog / Overview

Change how parametric functions and surfaces are placed to ensure they align with the x-axis and y-axis.
Change how the z-axis is placed to ensure it is where the x-axis crosses the y-axis.

## Motivation

Before these changes, parametric functions and surfaces were not aligned with the x-axis and y-axis when either did not include 0.  Similarly, the z-axis would also not be properly aligned when either axis did not include 0.

For further details, see [#1446 ](https://github.com/ManimCommunity/manim/issues/1446).  This PR fixes #1446 .

## Explanation for Changes

The changes made fix the placement of parametric functions/surfaces and the z-axis when either the x-axis or the y-axis do not include 0.

## Testing Status

pytest run via poetry completed successfully with no fails.

Additional tests run include visual inspections to ensure that everything aligned properly.  Visual tests include using various values in the below code to generate 2D graphs:

```py
from manim import *
import numpy as np

class ParametricPlot(Scene):
    def construct(self):
        axes = Axes(
                x_range=(5, 10, 1),
                y_range=(3, 8, 1)
                )

        trig_graph = ParametricFunction(lambda t: axes.c2p(t, 2 * np.sin(t) + 5), color=BLUE, t_range=(5, 10, 1))

        VGroup(axes, trig_graph).center()
        VGroup(axes, trig_graph).scale(0.5)
        self.add(axes, trig_graph)
```

Which resulted in the below graph:

ParametricPlot_PositiveNums

Next, `ThreeDAxes` was checked using this code:

```py
from manim import *
import numpy as np

class ThreeDSurfacePlot(ThreeDScene):
    def construct(self):
        resolution_fa = 42
        self.set_camera_orientation(phi=75 * DEGREES, theta=10 * DEGREES)

        axes = ThreeDAxes(
                x_range=(2, 7, 1),
                y_range=(3, 8, 1),
                z_range=(-1, 1, 0.5)
                )

        def param_surface(u, v):
            x = u
            y = v
            z = np.sin(x) * np.cos(y)
            return z

        surface_plane = ParametricSurface(
            lambda u, v: axes.c2p(u, v, param_surface(u, v)),
            resolution=(resolution_fa, resolution_fa),
            v_min=3,
            v_max=8,
            u_min=2,
            u_max=7,
        )

        surface_plane.set_style(fill_opacity=1, stroke_color=GREEN)
        surface_plane.set_fill_by_checkerboard(ORANGE, BLUE, opacity=0.5)

        self.add(axes, surface_plane)

        self.begin_ambient_camera_rotation(rate=-0.3)
        self.wait(10)
```

Which generated this graph:

https://user-images.githubusercontent.com/79891353/117085355-5c96f600-acfe-11eb-8d93-c4db20820911.mp4

## Checklist
- [X] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [X] I have written a descriptive PR title (see top of PR template for examples)
- [X] I have written a changelog entry for the PR or deem it unnecessary
- [ ] My new functions/classes either have a docstring or are private
- [ ] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by hydrobeam** [comment]
Thanks for solving this bug! Although, I think that since the bounds-checking structure is showing up multiple throughout `CoordinateSystem`, it should be written as a function instead now.

**Comment by alembcke** [comment]
That's a pretty good idea, as there are three places using the same logic.  How about something like:

```py
def origin_shift(self, x_range):
    if x_range[0] > 0:
        return x_range[0]
    if x_range[1] < 0:
        return x_range[1]
    else:
        return 0
```

Then change the code which shifts things to:

```py
axis.shift(axis.x_axis.number_to_point(origin_shift(self.x_range)))
```

Or something along those lines, depending on circumstances.  Is that what you were envisaging?  If so, then I will make the change.

**Comment by hydrobeam** [comment]
> Or something along those lines, depending on circumstances. Is that what you were envisaging? If so, then I will make the change.

That's essentially what I had in mind! I'm currently working on porting over `GraphScene`'s methods so this will also be useful there. However, to keep the function independent of which axis its based on, I would rename `x_range` to `axis_range`.

**Comment by alembcke** [comment]
@hydrobeam , I have made the requested changes, rerun pytest and visual inspections and all looks good.

File: manim/mobject/coordinate_systems.py [hunk 1]
```diff
@@ -367,6 +362,21 @@ def construct(self):
...
         return line_graph
 
+    def origin_shift(self, axis_range):
+        """Shifts graph mobjects to compensate when 0 is not on the axis.
+
+        Parameters
+        ----------
+        axis_range : Union[:class:`list`, :class:`numpy.ndarray`]
+            The range of the axis : `(x_min, x_max, x_step)`.
+        """
```
Comment (hydrobeam) [inline manim/mobject/coordinate_systems.py:372]: Typing + some minor corrections. I believe `List` might have to be imported

```suggestion
    def origin_shift(self, axis_range:List[float]):
        """Shifts graph mobjects to compensate when 0 is not on the axis.

        Parameters
        ----------
        axis_range
            The range of the axis : ``(x_min, x_max, x_step)``.
        """
```

**Review by hydrobeam — APPROVED** [review]

**Review by behackl — APPROVED** [review]
The changes look good to me, thank you for your contribution! 

@hydrobeam, can this be merged? I've just resolved the merge conflict in the imports.

**Comment by hydrobeam** [comment]
Looks good to me! Maybe we can implement some tests later so this doesn't happen again, but this PR is good to go.

**Comment by behackl** [comment]
> Looks good to me! Maybe we can implement some tests later so this doesn't happen again, but this PR is good to go.

I have added one test for checking the shift. Could you cross-review? If you are satisfied and if the pipeline passes, I'd like to get this merged. :-)

File: manim/mobject/coordinate_systems.py [hunk 2]
```diff
@@ -367,6 +362,21 @@ def construct(self):
...
         return line_graph
 
+    def origin_shift(self, axis_range: List[float]):
+        """Shifts graph mobjects to compensate when 0 is not on the axis.
+
+        Parameters
+        ----------
+        axis_range
+            The range of the axis : ``(x_min, x_max, x_step)``.
+        """
```
Comment (hydrobeam) [inline manim/mobject/coordinate_systems.py:372]: Add a return value and convert to a static method
```suggestion

    @staticmethod
    def origin_shift(axis_range: List[float]) -> float:
        """Determines how to shift graph mobjects to compensate when 0 is not on the axis.

        Parameters
        ----------
        axis_range
            The range of the axis : ``(x_min, x_max, x_step)``.
```

**Comment by naveen521kk** [comment]
@hydrobeam you must squash while merging.

**Pull request merged**